File: implementations/swift/Sources/RCLParser/Lexer.swift
enum TokenType { case identifier, string, number, equal, comma, dot, `do`, end, lbracket, rbracket, eof }
struct Token { let type: TokenType; let value: String; let line: Int; let column: Int }
struct LexerState { let pos: Int; let line: Int; let col: Int }

final class Lexer {
    private let input: [Character]
    private var pos = 0
    private var line = 1
    private var col = 1

    init(_ text: String) { self.input = Array(text) }

    func snapshot() -> LexerState { .init(pos: pos, line: line, col: col) }
    func restore(_ st: LexerState) { pos = st.pos; line = st.line; col = st.col }

    func nextToken() throws -> Token {
        skip()
        if eof { return .init(type: .eof, value: "", line: line, column: col) }
        let (l, c, ch) = (line, col, cur)
        switch ch {
        case "=": adv(); return .init(type: .equal, value: "=", line: l, column: c)
        case ",": adv(); return .init(type: .comma, value: ",", line: l, column: c)
        case ".": adv(); return .init(type: .dot, value: ".", line: l, column: c)
        case "[": adv(); return .init(type: .lbracket, value: "[", line: l, column: c)
        case "]": adv(); return .init(type: .rbracket, value: "]", line: l, column: c)
        case "\"": return try readString(line: l, column: c)
        default:
            if ch.isNumber || (ch == "-" && peek.isNumber) { return readNumber(line: l, column: c) }
            if ch.isLetter || ch == "_" { return readIdentifier(line: l, column: c) }
            throw ParseError("unexpected character", line: l, column: c)
        }
    }

    private func readString(line: Int, column: Int) throws -> Token {
        adv(); var out = ""
        while !eof && cur != "\"" {
            if cur == "\\" {
                adv(); if eof { throw ParseError("unterminated escape", line: line, column: column) }
                switch cur {
                case "\"": out.append("\"")
                case "n": out.append("\n")
                case "t": out.append("\t")
                case "\\": out.append("\\")
                default: throw ParseError("invalid escape", line: self.line, column: self.col)
                }
                adv()
            } else { out.append(cur); adv() }
        }
        if eof { throw ParseError("unterminated string", line: line, column: column) }
        adv()
        return .init(type: .string, value: out, line: line, column: column)
    }

    private func readNumber(line: Int, column: Int) -> Token {
        var out = ""
        if cur == "-" { out.append("-"); adv() }
        while !eof && cur.isNumber { out.append(cur); adv() }
        if !eof && cur == "." && peek.isNumber { out.append("."); adv(); while !eof && cur.isNumber { out.append(cur); adv() } }
        return .init(type: .number, value: out, line: line, column: column)
    }

    private func readIdentifier(line: Int, column: Int) -> Token {
        var out = ""
        while !eof && (cur.isLetter || cur.isNumber || cur == "_") { out.append(cur); adv() }
        if out == "do" { return .init(type: .do, value: out, line: line, column: column) }
        if out == "end" { return .init(type: .end, value: out, line: line, column: column) }
        return .init(type: .identifier, value: out, line: line, column: column)
    }

    private func skip() {
        while !eof {
            if cur.isWhitespace { adv() }
            else if cur == "#" { while !eof && cur != "\n" { adv() } }
            else { break }
        }
    }

    private var eof: Bool { pos >= input.count }
    private var cur: Character { input[pos] }
    private var peek: Character { pos + 1 < input.count ? input[pos + 1] : "\0" }
    private func adv() { if eof { return }; if cur == "\n" { line += 1; col = 1 } else { col += 1 }; pos += 1 }
}

File: packages/typescript/src/lexer.ts
import { ParserException } from "./error.js";
import { Token, TokenType } from "./token.js";

type LexerState = { pos: number; line: number; column: number };

export class Lexer {
  private pos = 0;
  private line = 1;
  private column = 1;

  constructor(private readonly input: string) {}

  snapshot(): LexerState {
    return { pos: this.pos, line: this.line, column: this.column };
  }

  restore(state: LexerState): void {
    this.pos = state.pos;
    this.line = state.line;
    this.column = state.column;
  }

  nextToken(): Token {
    this.skipWhitespaceAndComments();
    if (this.eof()) return this.mk(TokenType.EOF, "", this.line, this.column);

    const line = this.line;
    const col = this.column;
    const ch = this.current();

    if (ch === "=") return this.single(TokenType.Equal, "=", line, col);
    if (ch === ",") return this.single(TokenType.Comma, ",", line, col);
    if (ch === ".") return this.single(TokenType.Dot, ".", line, col);
    if (ch === "[") return this.single(TokenType.LBracket, "[", line, col);
    if (ch === "]") return this.single(TokenType.RBracket, "]", line, col);
    if (ch === '"') return this.readString(line, col);
    if (this.isNumberStart(ch)) return this.readNumber(line, col);
    if (this.isIdentStart(ch)) return this.readIdent(line, col);

    throw new ParserException(`unexpected character '${ch}'`, line, col);
  }

  private readString(line: number, col: number): Token {
    this.advance();
    let out = "";
    while (!this.eof() && this.current() !== '"') {
      if (this.current() === "\\") {
        this.advance();
        if (this.eof()) throw new ParserException("unterminated escape sequence", line, col);
        const esc = this.current();
        if (esc === '"') out += '"';
        else if (esc === "n") out += "\n";
        else if (esc === "t") out += "\t";
        else if (esc === "\\") out += "\\";
        else throw new ParserException(`invalid escape \\${esc}`, this.line, this.column);
        this.advance();
      } else {
        out += this.current();
        this.advance();
      }
    }
    if (this.eof()) throw new ParserException("unterminated string", line, col);
    this.advance();
    return this.mk(TokenType.String, out, line, col);
  }

  private readNumber(line: number, col: number): Token {
    let out = "";
    if (this.current() === "-") {
      out += "-";
      this.advance();
    }
    while (!this.eof() && /[0-9]/.test(this.current())) {
      out += this.current();
      this.advance();
    }
    if (!this.eof() && this.current() === "." && /[0-9]/.test(this.peek() ?? "")) {
      out += ".";
      this.advance();
      while (!this.eof() && /[0-9]/.test(this.current())) {
        out += this.current();
        this.advance();
      }
    }
    return this.mk(TokenType.Number, out, line, col);
  }

  private readIdent(line: number, col: number): Token {
    let out = "";
    while (!this.eof() && this.isIdentPart(this.current())) {
      out += this.current();
      this.advance();
    }
    if (out === "do") return this.mk(TokenType.Do, out, line, col);
    if (out === "end") return this.mk(TokenType.End, out, line, col);
    return this.mk(TokenType.Identifier, out, line, col);
  }

  private skipWhitespaceAndComments(): void {
    while (!this.eof()) {
      const ch = this.current();
      if (/\s/.test(ch)) this.advance();
      else if (ch === "#") this.skipToEol(); else break;
    }
  }

  private skipToEol(): void {
    while (!this.eof() && this.current() !== "\n") this.advance();
  }

  private single(type: TokenType, value: string, line: number, col: number): Token {
    this.advance();
    return this.mk(type, value, line, col);
  }

  private mk(type: TokenType, value: string, line: number, column: number): Token {
    return { type, value, line, column };
  }

  private isIdentStart(ch: string): boolean { return /[A-Za-z_]/.test(ch); }
  private isIdentPart(ch: string): boolean { return /[A-Za-z0-9_]/.test(ch); }
  private isNumberStart(ch: string): boolean {
    return /[0-9]/.test(ch) || (ch === "-" && /[0-9]/.test(this.peek() ?? ""));
  }

  private advance(): void {
    if (this.eof()) return;
    if (this.current() === "\n") { this.line += 1; this.column = 1; }
    else this.column += 1;
    this.pos += 1;
  }
  private current(): string { return this.input[this.pos] ?? ""; }
  private peek(): string | undefined { return this.input[this.pos + 1]; }
  private eof(): boolean { return this.pos >= this.input.length; }
}
